### ISSUE TITLE
Node debug config provider doesn't work?

### ISSUE THREAD

**Issue by roblourens** [Microsoft/vscode#47779]
- Start with --disable-extensions
- Open the smoketest repo
- "Select environment" popup appears, and is empty
- Press esc
- Opens settings for some reason

If I start with extensions enabled, I see other extensions but not node.

Also, if I click the "configure or open launch.json" button, it just opens the settings editor. Related issue or different?

**Comment by weinand** [comment]
@roblourens I cannot reproduce this with Stable or Insiders.

**Comment by roblourens** [comment]
I can repro in both, and on a Linux VM.

If I reload the window then trigger/esc multiple times, I sometimes see Node.js appear after a few iterations. I think it's showing the popup without wiating for the extension to finish activating.

**Comment by isidorn** [comment]
@roblourens I cannot reproduce this on my mac, I nicely get offered Node and launch.json always nicely gets opened. If I understand corerctly you see this both on linux and on mac.

Can you still see it with latest vscode insiders?
If yes, put a bp here and check the candidates array https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/debug/electron-browser/debugConfigurationManager.ts#L516
Are you in some multi root repo? Can you repro if you use a differnt user-data-dir?

**Comment by roblourens** [comment]
Still happens for me.

Latest insiders with extensions disabled:

- Pressing F5 with no files open
  - Empty "Select Environment" popup appears.
  - I press esc
  - Settings editor appears
  - If I force the node debug extension to activate (Like by running the Toggle Auto Attach) command, it works
  - Debugging, I see that `this.debuggers` has the debuggers but `hasConfigurationProvider` is false so `candidates` stays empty
- Pressing F5 with app.js open
  - Debugging, I see it pick node-debug from the language
  - Error popup "Cannot find a program to debug"
  - Settings editor opens

With fresh user-data-dir and extensions disabled:

- Pressing F5 with no files open
  - Empty "Select Environment" popup appears.
  - I press esc
  - "Select Environment" appears a second time with Node.js
  - Debugging it, the second time node-debug does have `hasConfigurationProvider: true`
- Pressing F5 with app.js open
  - Works fine

Seems that there's partly something wrong in my userDataDir but it's not perfect with a fresh one either.

I've had a default launch config in my user settings but I comment it out and reload before reproing this.

**Comment by roblourens** [comment]
I think that the launch config in user settings is actually persisted as the current one even though it doesn't exist anymore. `launch` of "user settings" is passed to `startDebugging`. That's why it keeps opening the settings editor.

But I don't see how the extensions are supposed to be activated for onDebugInitialConfigurations before startDebugging gets to calling guessDebugger.

**Comment by roblourens** [comment]
But onDebugResolve happens after the call to guessDebugger in my first scenario. If I force node-debug to activate, it does work, so I think that's the problem. Can we debug over a call?

**Comment by isidorn** [comment]
@roblourens Yeah, let's debug over a call. How about tomorrow morning your time / evening my time?
Ping me on slack

**Comment by roblourens** [comment]
Sure. I agree that there are two issues. The second one is also showing up in another way, I keep seeing the debug config dropdown showing configs from different vscode windows. @octref saw the same thing. But I can't repro it consistently and it goes away after a reload.

**Comment by roblourens** [comment]
Another way I'm seeing it:

- Run the "Launch VS Code"
- Using the debug quickopen menu (`debug ...`), run the "Attach to Extension Host" config
- The debug dropdown still shows "Launch VS Code", and if I press F5 later, it's actually running "Attach to Extension Host"
- And if I open the quickopen again later, it shows the correct config selected initially.

![image](https://user-images.githubusercontent.com/323878/39215952-b598e3d2-47ce-11e8-9f6c-a3c64a2b4238.png)

Sorry if I'm mixing more unrelated issues in here but it seems relevant and these steps are simpler.

**Comment by isidorn** [comment]
@roblourens your latest comment with steps is fixed in latest vscode insiders. I am investigating into the original issue...

**Comment by isidorn** [comment]
To verify, check that the node debug gets activated (you get it in the quick open) in the following situations each right after a reload:
1. Open debug config (no launch.json present)
2. Start debugging (no launch.json present)
3. Also with launch.json there
